src/drivers/scsi/scsi.c

```diff
@@ -69,6 +59,105 @@ int scsi_do_cmd(struct scsi_dev *dev, struct scsi_cmd *cmd) {
 	return scsi_cmd(dev, scmd, cmd->scmd_len, cmd->scmd_obuf, cmd->scmd_olen);
 }
 
+static int scsi_send_inquiry(struct scsi_dev *dev) {
+	struct scsi_cmd cmd = scsi_cmd_template_inquiry;
+	struct scsi_data_inquiry *data;
+
+	log_debug("");
+
+	cmd.scmd_obuf = dev->scsi_data_scratchpad;
+	cmd.scmd_olen = USB_SCSI_SCRATCHPAD_LEN;
+	if (scsi_do_cmd(dev, &cmd) < 0) {
+		return -1;
+	}
+
+	data = (struct scsi_data_inquiry *) dev->scsi_data_scratchpad;
+	if ((data->dinq_devtype & SCSI_INQIRY_DEVTYPE_MASK)
+			!= SCSI_INQIRY_DEVTYPE_BLK) {
+		return -1;
+	}
+	return 0;
+}
+
+static int scsi_send_test_unit(struct scsi_dev *dev) {
+	struct scsi_cmd cmd = scsi_cmd_template_test_unit;
+
+	log_debug("");
+
+	cmd.scmd_olen = 0;
+	return scsi_do_cmd(dev, &cmd);
+}
+
+static int scsi_send_sense(struct scsi_dev *dev) {
+	struct scsi_cmd cmd = scsi_cmd_template_sense;
+	struct scsi_data_sense *data;
+	uint8_t acode;
+
+	log_debug("");
+
+	cmd.scmd_obuf = dev->scsi_data_scratchpad;
+	cmd.scmd_olen = sizeof(struct scsi_data_sense);
+	if (scsi_do_cmd(dev, &cmd) < 0) {
+		return -1;
+	}
+
+	data = (struct scsi_data_sense *) dev->scsi_data_scratchpad;
+	acode = data->dsns_additional_code;
+	if (!(acode == 0x28 || acode == 0x29)) {
```

Comment:
May be using MACRO for code definition?

Reply:
Uhh, it's difficult to say what is a correct check for this. I was looking into SCSI specification and found that for example value `0x29` (look at the first column) also requires to check scsi code qualifier (the second column), so I cannot even define such a MACRO, because it may be a combination of two codes, who knows...

```
29 00 Power On, Reset, Or Bus Device Reset Occurred 6
29 01 Power-On Reset Occurred 6
29 02 SCSI Bus Reset Occurred 6
29 03 Bus Device Reset Function Occurred 6
29 04 Internal Reset Occurred 6
29 05 Transceiver Mode Changed To Single-Ended 6
29 06 Transceiver Mode Changed To LVD 6
29 07 Write Log Dump data to disk successful OR IT Nexus Loss 6
29 08 Write Log Dump data to disk fail 6
29 09 Write Log Dump Entry information fail 6
29 0A Reserved disk space is full 6
29 0B SDBP test service contained an error, examine status packet(s) for detail
```